crates/tracing/src/lib.rs

```diff
@@ -0,0 +1,279 @@
+pub mod context;
+
+use std::io::{IsTerminal, StderrLock, Write};
+
+#[cfg(feature = "otel")]
+use opentelemetry::sdk::{
+    trace::{self, RandomIdGenerator, Sampler, Tracer},
+    Resource,
+};
+#[cfg(feature = "otel")]
+use opentelemetry::trace::TraceError;
+#[cfg(feature = "otel")]
+use opentelemetry_otlp::{Protocol, WithExportConfig};
+use tracing::{Event, Subscriber};
+use tracing_subscriber::fmt::format::{DefaultFields, JsonFields};
+use tracing_subscriber::fmt::{
+    format::{Format, Full, Json, Writer},
+    time::SystemTime,
+    FmtContext, FormatEvent, FormatFields,
+};
+use tracing_subscriber::{
+    filter::LevelFilter,
+    layer::{Layered, SubscriberExt},
+    registry::LookupSpan,
+    EnvFilter, Layer, Registry,
+};
+
+use wasmcloud_core::{logging::Level, OtelConfig};
+
+struct LockedWriter<'a> {
+    stderr: StderrLock<'a>,
+}
+
+impl<'a> LockedWriter<'a> {
+    fn new() -> Self {
+        LockedWriter {
+            stderr: STDERR.lock(),
+        }
+    }
+}
+
+impl<'a> Write for LockedWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.stderr.write(buf)
+    }
+
+    /// DIRTY HACK: when flushing, write a carriage return so the output is clean and then flush
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.stderr.write_all(&[13])?;
+        self.stderr.flush()
+    }
+}
+
+impl<'a> Drop for LockedWriter<'a> {
+    fn drop(&mut self) {
+        let _ = self.flush();
+    }
+}
+
+lazy_static::lazy_static! {
+    static ref STDERR: std::io::Stderr = std::io::stderr();
+}
```

Comment:
Can we migrate to `once_cell` please? I'm fine with adding a TODO here and in `Cargo.toml`